### PR TITLE
(BQ Python) Perform job waits in finish_bundle to allow BQ streaming writes with large batch loads

### LINKED ISSUE #23104: [Bug]: WriteToBigQuery with file_loads and dynamic table destination doesn't load after first File Load
### What happened?

I was able to reproduce a bug in the WriteToBigQuery function in the apache beam source code for Dataflow Streaming and the File Loads setting.  This issue is primarily caused by the impulse nodes feeding into the WaitForBQJobs task used in four separate nodes in the pipeline. 

This bug manifests when using file loads to write data to different tables, set via a lambda reading the row inputs. The issue occurs after the first file load, where the impulses (which should either be periodic, or through a different method) no longer fire again, resulting in the side inputs into the nodes that load to a temp table, check if the load is complete, load a schema change to the intended table, check if that change is complete, copy from the temp table to the intended table, check that is complete, then remove the original temp table. 

The use of the single impulse makes sure that this set of work gets executed once, but after it is executed once, and more data comes in, none of that data gets actioned.  

I reported this issue directly to the Customer Support at the Dataflow team at GCP, but after not seeing it being actioned, I'm reporting it here. 

[Here is the Google Doc I created for them that walks through reproduction of the bug.](https://docs.google.com/document/d/1uIM5JVq0dAh2uDB0HfzQN7PS60U8TsJnalvAvfkfcnM/edit?usp=sharing)

This has been reproduced on both Apache beam 2.32 and 2.40, running on dataflow.

### Issue Priority

Priority: 2

### Issue Component

Component: io-py-gcp

### PR DESCRIPTION
**Pull request by ahmedabu98**
Streaming FILE_LOADS is currently not working with large loads. After the first load, it lags because `WaitForBQJobs` is only triggered once with `beam.Create([None])`. These changes move the wait into the finish_bundle of its respective step. 

Note: Streaming with small loads is done in one step and thus was not getting stuck.

Also adding tests that more accurately determine if streaming with batch loads is working.

Needs #23011 to run tests correctly

Fixes #23104

### CODE REVIEW

**Comment by ahmedabu98** [comment]
Waiting for #23011

**Comment by ahmedabu98** [comment]
Run Python 3.8 PostCommit

File: sdks/python/apache_beam/io/gcp/bigquery_file_loads.py [hunk 1]
```diff
@@ -353,9 +356,10 @@ def __init__(
     self._step_name = step_name
     self._load_job_project_id = load_job_project_id
 
-  def setup(self):
-    self._bq_wrapper = bigquery_tools.BigQueryWrapper(client=self._test_client)
+  def start_bundle(self):
+    self.bq_wrapper = bigquery_tools.BigQueryWrapper(client=self._test_client)
```
Comment (johnjcasey) [inline sdks/python/apache_beam/io/gcp/bigquery_file_loads.py:360]: why the change between _bq_wrapper and bq_wrapper?

File: sdks/python/apache_beam/io/gcp/bigquery_file_loads.py [hunk 2]
```diff
@@ -430,20 +434,36 @@ def process(self, element, schema_mod_job_name_prefix):
...
+  def finish_bundle(self):
+    # Unlike the other steps, schema update is not always necessary.
+    # In that case, return a None value to avoid blocking in streaming context.
+    # Otherwise, the streaming pipeline would get stuck waiting for the
+    # TriggerCopyJobs side-input.
+    if not self.pending_jobs:
+      return [GlobalWindows.windowed_value(None)]
+
+    for windowed_value in self.pending_jobs:
+      job_ref = windowed_value.value[1]
```
Comment (johnjcasey) [inline sdks/python/apache_beam/io/gcp/bigquery_file_loads.py:463]: Do we want to loop over these waits? that seems somewhat out of pattern.

**Review by johnjcasey — CHANGES_REQUESTED** [review]

File: sdks/python/apache_beam/io/gcp/bigquery_file_loads.py (hunk 1 above)
Comment (ahmedabu98) [reply sdks/python/apache_beam/io/gcp/bigquery_file_loads.py:360]: Everywhere else in the file it's bq_wrapper, so I wanted to keep it consistent

File: sdks/python/apache_beam/io/gcp/bigquery_file_loads.py (hunk 2 above)
Comment (ahmedabu98) [reply sdks/python/apache_beam/io/gcp/bigquery_file_loads.py:463]: Hmmm I'm not sure I follow, what stands out as weird to you?

FYI Java implementation is somewhat like this. Jobs are submitted and accumulated as they are pending. Then they are[ looped over and waited for in finishBundle](https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java#L344).

**Comment by ahmedabu98** [comment]
Run Python 3.8 PostCommit

**Comment by ahmedabu98** [comment]
Run Python 3.8 PostCommit

**Comment by ahmedabu98** [comment]
Run Python PreCommit

**Comment by ahmedabu98** [comment]
Run Python PreCommit

**Comment by ahmedabu98** [comment]
Run Python 3.8 PostCommit

**Comment by ahmedabu98** [comment]
Run Python 3.8 PostCommit

File: sdks/python/apache_beam/io/gcp/bigquery_file_loads.py (hunk 2 above)
Comment (johnjcasey) [reply sdks/python/apache_beam/io/gcp/bigquery_file_loads.py:465]: In that one, we add all of the waits to a manager, and then we wait for the manager. In this, we wait directly. Ideally we can do that pattern here as well

**Review by johnjcasey — CHANGES_REQUESTED** [review]

File: sdks/python/apache_beam/io/gcp/bigquery_file_loads.py (hunk 2 above)
Comment (ahmedabu98) [reply sdks/python/apache_beam/io/gcp/bigquery_file_loads.py:465]: Waiting for jobs in a for loop is how the [current implementation](https://github.com/apache/beam/blob/2d4f61c93250fb41112a2535394e7328bc1fdf95/sdks/python/apache_beam/io/gcp/bigquery_file_loads.py#L757) does it as well. Looks like this is because there's no [PendingJobManager](https://github.com/apache/beam/blob/935eeadd45a6100c68243fc5b266b20d26164173/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryHelpers.java#L91) equivalent for the Python connector.

Comment (ahmedabu98) [reply sdks/python/apache_beam/io/gcp/bigquery_file_loads.py:465]: The benefit of the PendingJobManager's [wait implementation](https://github.com/apache/beam/blob/935eeadd45a6100c68243fc5b266b20d26164173/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryHelpers.java#L127) is that it waits for the jobs in parallel and checks on each of them. This way, one slow job does not block us from checking (and logging) the state of the other jobs.

This is lost in the Python connector's wait implementation because we check the state of the jobs sequentially. Ultimately, however, it doesn't change the big picture behavior because file loads are atomic, ie. if one job fails, all the jobs in the bundle fail. We don't release jobs to the next step until all the jobs are finished waiting.

Comment (ahmedabu98) [reply sdks/python/apache_beam/io/gcp/bigquery_file_loads.py:465]: If we decide a job manager for this connector would be helpful, perhaps it could be added in a separate PR?

Comment (johnjcasey) [reply sdks/python/apache_beam/io/gcp/bigquery_file_loads.py:465]: Sounds good. Can you write that up as an improvement?

this looks good to me otherwise
